overlord/exportstate/exportstate.go

```diff
@@ -0,0 +1,450 @@
+		exports = make(map[string]*json.RawMessage)
+	}
+	key := stateMapKey(instanceName, rev)
+	if m == nil {
+		delete(exports, key)
+	} else {
+		data, err := json.Marshal(m)
+		if err != nil {
+			panic("internal error: cannot marshal export manifest: " + err.Error())
+		}
+		raw := json.RawMessage(data)
+		exports[key] = &raw
+	}
+	st.Set("exports", exports)
+}
+
+// Get recalls export manifest of a particular snap revision.
+func Get(st *state.State, instanceName string, rev snap.Revision, m *Manifest) error {
+	*m = Manifest{}
+
+	var exports map[string]*json.RawMessage
+	if err := st.Get("exports", &exports); err != nil {
+		// This can return state.ErrNoState
+		return err
+	}
+	key := stateMapKey(instanceName, rev)
+	raw, ok := exports[key]
+	if !ok {
+		return state.ErrNoState
+	}
+	// XXX: do we need the address?
+	if err := json.Unmarshal([]byte(*raw), &m); err != nil {
+		return fmt.Errorf("cannot unmarshal export manifest: %v", err)
+	}
+	return nil
+}
+
+// currentSymlinkPath returns the path of the current subkey symlink for given primaryKey.
+func currentSubKeySymlinkPath(primaryKey string) string {
+	return filepath.Join(ExportDir, primaryKey, "current")
+}
+
+// setCurrentSubKey replaces the "current" symlink for the given primary key to
+// point to the given subKey. Appropriate subKey can be computed by
+// subKeyForSnap.
+//
+// If the symbolic link cannot be created because the export directory does not
+// exist no error is reported. This is because this function is most often
+// called from link-snap where it runs unconditionally but most snaps do not
+// have any content to export and the symlink would be dangling.
+func setCurrentSubKey(primaryKey, subKey string) error {
+	pathName := currentSubKeySymlinkPath(primaryKey)
+	if err := osutil.AtomicSymlink(subKey, pathName); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("cannot set current subkey of %q to %q: %v", primaryKey, subKey, err)
+	}
+	return nil
+}
+
+// removeCurrentSubKey removes the "current" symlink for the given primary key.
+func removeCurrentSubKey(primaryKey string) error {
+	if err := os.Remove(currentSubKeySymlinkPath(primaryKey)); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("cannot remove current subkey of %q: %v", primaryKey, err)
+	}
+	return nil
+}
+
+// ManifestKeys returns the (primaryKey, subKey) tuple to use as the current
+// provider of all the export sets of a given snap. The returned subKey may be
+// empty, indicating that given snap has no current revision.
+func ManifestKeys(st *state.State, instanceName string) (primaryKey string, subKey string, err error) {
+	switch instanceName {
+	case "core", "snapd":
+		primaryKey, subKey, err = manifestKeysForSnapdOrCore(st)
+		if err != nil {
+			return "", "", err
+		}
+	default:
+		info, err := snapstateCurrentInfo(st, instanceName)
+		if _, ok := err.(*snap.NotInstalledError); err != nil && !ok {
+			return "", "", err
+		}
+		if info == nil || info.Broken != "" {
```

Comment:
Why aren't we erroring out if broken?

Reply:
To be clear I'm not sure we should error out, it is mostly a question. Perhaps it deserves a comment in the code.

Reply:
We're not erroring on broken, we just don't want to attempt to use them. This came out of the snapd failover test.